source/slang/syntax.cpp

```diff
@@ -2061,4 +1918,53 @@ void Type::accept(IValVisitor* visitor, void* extra)
         }*/
         return newSubst;
     }
+    bool SubstitutionSet::Equals(SubstitutionSet substSet) const
+    {
+        if (genericSubstitutions)
+        {
+            if (!genericSubstitutions->Equals(substSet.genericSubstitutions))
+                return false;
+        }
+        else
+        {
+            if (substSet.genericSubstitutions)
+                return false;
+        }
+        if (thisTypeSubstitution)
+        {
+            if (!thisTypeSubstitution->Equals(substSet.thisTypeSubstitution))
+                return false;
+        }
+        else
+        {
+            if (substSet.thisTypeSubstitution && substSet.thisTypeSubstitution->sourceType != nullptr)
+                return false;
+        }
+        return true;
+    }
+    SubstitutionSet SubstitutionSet::substituteImpl(SubstitutionSet subst, int * ioDiff)
+    {
+        SubstitutionSet rs;
+        if (genericSubstitutions)
+            rs.genericSubstitutions = genericSubstitutions->SubstituteImpl(subst, ioDiff).As<GenericSubstitution>();
+        if (globalGenParamSubstitutions)
+            rs.globalGenParamSubstitutions = globalGenParamSubstitutions->SubstituteImpl(subst, ioDiff).As<GlobalGenericParamSubstitution>();
+        if (thisTypeSubstitution)
+            rs.thisTypeSubstitution = thisTypeSubstitution->SubstituteImpl(subst, ioDiff).As<ThisTypeSubstitution>();
+        else
+            rs.thisTypeSubstitution = subst.thisTypeSubstitution;
```

Comment:
I suspect we shouldn't have this line. If the `this` substitution doesn't have a `ThisTypeSubst`, then it doesn't make sense to copy the one from `subst`. This would be like saying "if the left hand side doesn't have any generics, then copy the generic substitutions form the other side."